src/program/latex/diagram/mod.rs

```diff
@@ -0,0 +1,176 @@
+                self.wires.insert(qubit, i);
+            }
+        } else {
+            for (i, qubit) in qubits.iter().enumerate() {
+                self.wires.insert(*qubit, i);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Applies a gate from an instruction to the associated wires in the
+    /// circuit of this diagram.
+    ///
+    /// # Arguments
+    /// `gate` - the Gate of the Instruction from `to_latex`.
+    pub(crate) fn add_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
+        let mut cells: Vec<QuantikzCell> = vec![QuantikzCell::Empty; self.wires.len()];
+
+        let quantikz_gate = cell::QuantikzGate::try_from(gate.clone())?;
+
+        if gate.qubits.len() - quantikz_gate.ctrl_count > 1 {
+            return Err(LatexGenError::UnsupportedGate {
+                gate: gate.name.to_string(),
+            });
+        }
+
+        // parameterized non-PHASE gates are unsupported
+        if quantikz_gate.parameter.is_some() && !gate.name.contains("PHASE") {
+            return Err(LatexGenError::UnsupportedGate {
+                gate: gate.name.clone(),
+            });
+        }
+
+        let mut qubit_iterator = gate
+            .qubits
+            .iter()
+            .map(|q| {
+                q.clone()
+                    .into_fixed()
+                    .map_err(|_| LatexGenError::UnsupportedQubit)
+            })
+            .rev();
+
+        let target_qubit = qubit_iterator
+            .next()
+            .ok_or(LatexGenError::RequiresQubits)??;
+
+        let target_index = *self
+            .wires
+            .get(&target_qubit)
+            .ok_or(LatexGenError::QubitNotFound(target_qubit))?;
+
+        cells[target_index] = QuantikzCell::Gate(quantikz_gate);
+
+        for qubit in qubit_iterator {
+            let column_index = *self
+                .wires
+                .get(&qubit?)
+                .ok_or(LatexGenError::QubitNotFound(target_qubit))?;
+            cells[column_index] = QuantikzCell::Control(target_index as i64 - column_index as i64)
+        }
+
+        self.columns.push(DiagramColumn { cells });
+
+        Ok(())
+    }
+}
+
+impl fmt::Display for Diagram {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let last = self.wires.keys().last().unwrap_or(&0);
+        let cell_settings = CellSettings::from(self.settings);
+
+        for (qubit, cell_index) in &self.wires {
+            if self.settings.label_qubit_lines {
+                write!(f, "{}", RenderCommand::LeftWireLabel(*qubit))?;
+            } else {
+                write!(f, "{}", RenderCommand::Qw)?;
+            }
+
+            for column in &self.columns {
+                let command = &column.cells[*cell_index];
+                write!(
+                    f,
+                    "{}{}",
+                    &RenderCommand::Separate,
+                    command.to_latex(&cell_settings).map_err(|_| fmt::Error)?
```

Comment:
I'm not a fan of erasing errors like this. See my note above about having a `display_latex` function, which avoids this.

From the docs for `std::fmt::Error`:

> This type does not support transmission of an error other than that an error occurred. Any extra information must be arranged to be transmitted through some other means.

IMO, the only error that should come from `Display` is to do with the actual writing of text, not validating the thing to write